docs/src/design/decoder/main.md

```diff
@@ -0,0 +1,517 @@
+8. Two additional registers (not shown) used primarily for constraint degree reduction.
+
+### Program block hashing
+To compute hashes of program blocks, the decoder relies on the hash co-processor. Specifically, the decoder needs to perform two types of hashing operations:
+
+1. A simple 2-to-1 hash, where we provide a sequence of $8$ field elements, and get back $4$ field elements representing the result. Computing such a hash requires $8$ rows in the hash co-processor.
+2. A sequential hash of $n$ elements. Computing such a hash requires multiple absorption steps, and at each step $8$ field elements are absorbed into the hasher. Thus, commuting a sequential hash of $n$ elements requires $\lceil {n/8} \rceil$ rows in the hash co-processor. At the end, we also get $4$ field elements representing the result.
+
+We denote the running product column used to keep track of hash co-processor state as $p_0$. To make hashing requests to the hash co-processor and to read the results from it, we will need to divide out relevant values from this column as described below.
+
+#### Simple 2-to-1 hash
+To initiate a 2-to-1 hash of $8$ elements ($v_0, ..., v_7$) we need to divide $p_0$ by the following value:
+
+$$
+\alpha_0 + \alpha_1 \cdot m_{bp} + \alpha_2 \cdot r + \alpha_4 \cdot 8 + \sum_{i=0}^7 (\alpha_{i+8} \cdot v_i)
+$$
+
+where:
+* $m_{bp}$ is a label indicating beginning of a new permutation.
+* $r$ is the address of the row at which the hashing begins.
+* $\alpha_4 \cdot 8$ indicates that we are hashing a total of $8$ elements.
+
+To read the $4$-element result ($u_0, ..., u_3$), we need to divide $p_0$ by the following value:
+
+$$
+\alpha_0 + \alpha_1 \cdot m_{hout} + \alpha_2 \cdot (r + 7) + \sum_{i=0}^3 (\alpha_{i+8} \cdot u_i)
+$$
+
+where:
+* $m_{hout}$ is a label indicating return of the hash value.
+* $r$ is the address of the row at which the hashing began.
+
+#### Sequential hash
+To initiate a sequential hash of $n$ elements ($v_0, ..., v_{n-1}$), we need to divide $p_0$ by the following value:
+
+$$
+\alpha_0 + \alpha_1 \cdot m_{bp} + \alpha_2 \cdot r + \alpha_4 \cdot n + \sum_{i=0}^7 (\alpha_{i+8} \cdot v_i)
+$$
+
+This also absorbs the first $8$ elements of the sequence into the hasher state. Then, to absorb the next sequence of $8$ elements (e.g., $v_8, ..., v_{15}$), we need to divide $p_0$ by the following value:
+
+$$
+\alpha_0 + \alpha_1 \cdot m_{abp} + \alpha_2 \cdot (r + 7) + \sum_{i=0}^7 (\alpha_{i+8} \cdot v_{i + 8})
+$$
+
+We can keep absorbing elements into the hasher in the similar manner until all elements have been absorbed. Then, to read the result (e.g., $u_0, ..., u_3$), we need to divide $p_0$ by the following value:
+
+$$
+\alpha_0 + \alpha_1 \cdot m_{hout} + \alpha_2 \cdot (r + \lceil n / 8 \rceil \cdot 8  - 1) + \sum_{i=0}^3 (\alpha_{i+8} \cdot u_i)
+$$
+
+Thus, for example, if $n = 16$, the result will of the hash will be available at hasher row $r + 15$.
```

Comment:
Given that we're using the ceiling of n/8, I think the example would be more useful if it weren't a multiple of 8.

Reply:
Good point! Updated!